core/modules/replica/WorkerProcessor.cc

```diff
@@ -690,47 +690,23 @@ void WorkerProcessor::_setServiceResponseInfo(
         WorkerRequest::Ptr const& request,
         ProtocolServiceResponseInfo* info) const {
 
-    if (
-        auto const ptr = dynamic_pointer_cast<WorkerReplicationRequest>(request)) {
-
-        info->set_replica_type(ProtocolReplicaRequestType::REPLICA_CREATE);
-        info->set_id(          ptr->id());
-        info->set_priority(    ptr->priority());
-        info->set_database(    ptr->database());
-        info->set_chunk(       ptr->chunk());
-        info->set_worker(      ptr->sourceWorker());
-
-    } else if (
-        auto const ptr = dynamic_pointer_cast<WorkerDeleteRequest>(request)) {
-
-        info->set_replica_type(ProtocolReplicaRequestType::REPLICA_DELETE);
-        info->set_id(          ptr->id());
-        info->set_priority(    ptr->priority());
-        info->set_database(    ptr->database());
-        info->set_chunk(       ptr->chunk());
-
-    } else if (
-        auto const ptr = dynamic_pointer_cast<WorkerFindRequest>(request)) {
-
-        info->set_replica_type(ProtocolReplicaRequestType::REPLICA_FIND);
-        info->set_id(          ptr->id());
-        info->set_priority(    ptr->priority());
-        info->set_database(    ptr->database());
-        info->set_chunk(       ptr->chunk());
-
-    } else if (
-        auto const ptr = dynamic_pointer_cast<WorkerFindAllRequest>(request)) {
-
-        info->set_replica_type(ProtocolReplicaRequestType::REPLICA_FIND_ALL);
-        info->set_id(          ptr->id());
-        info->set_priority(    ptr->priority());
-        info->set_database(    ptr->database());
-
+    if (nullptr != dynamic_pointer_cast<WorkerReplicationRequest>(request)) {
+        info->set_queued_type(ProtocolQueuedRequestType::REPLICA_CREATE);
+    } else if (nullptr != dynamic_pointer_cast<WorkerDeleteRequest>(request)) {
+        info->set_queued_type(ProtocolQueuedRequestType::REPLICA_DELETE);
+    } else if (nullptr != dynamic_pointer_cast<WorkerFindRequest>(request)) {
+        info->set_queued_type(ProtocolQueuedRequestType::REPLICA_FIND);
+    } else if (nullptr != dynamic_pointer_cast<WorkerFindAllRequest>(request)) {
+        info->set_queued_type(ProtocolQueuedRequestType::REPLICA_FIND_ALL);
+    } else if (nullptr != dynamic_pointer_cast<WorkerEchoRequest>(request)) {
+        info->set_queued_type(ProtocolQueuedRequestType::TEST_ECHO);
```

Comment:
all this dynamic_casting is not great I think. It would be good if there was a way to leverage the polymorphism to make a single call to get the proper ProtocolQueuedRequestType.

Reply:
You've hit right at the spot! The dynamic casting is a price to pay for my attempt to keep a hierarchy of transient classes (subclasses of `WorkerRequest`) to be completely independent of the types in the Protobuf definition file `replica/protocol.proto`. The only reason to have the ugly "staircase" of the `if` statements is to deduce a specific variant of this `enum` from th protocol definition file:
```
enum ProtocolQueuedRequestType {

    REPLICA_CREATE   = 0;    // replicate a chunk
    REPLICA_DELETE   = 1;    // delete an existing replica of a chunk
    REPLICA_FIND     = 2;    // find out if a replica is present, and report its state
    REPLICA_FIND_ALL = 3;    // find all replicas and report their states
    TEST_ECHO        = 4;    // test the worker-side framework
    SQL              = 5;    // execute an SQL statement against the worker's database
}
```
I could (in theory) introduce a method:
```
ProtocolQueuedRequestType WorkerRequest::requestType() const;
```
But this will add the direct dependency of the Worker request classes on the types generated by the Protobuf compiler.
Keeping transient classes separate from the protocol definitions has certain advantages...as well as problems. And, frankly, I'm still undecided which approach is the best. For now, I'd suggest keeping this code "as is". Otherwise, this code review will result in a significant amount of work. Let's address this issue in a separate refactoring ticket.

Reply:
I do not think we'll have more than one protocol supported (at a time) so direct dependency on protocol classes should be OK. Otherwise you'd need double dispatch to support multiple protocols.

Reply:
Even though I may agree with (both of) you, let's leave this topic for now. I will keep it in my _TO DO_ list. To tell the truth, refactoring the worker-side Replication Framework would require an effort comparable with the one of the current ticket.